src/Config/ObjectModel/DataSource.cs

```diff
@@ -0,0 +1,68 @@
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.DataApiBuilder.Config.NamingPolicies;
+
+namespace Azure.DataApiBuilder.Config.ObjectModel;
+
+/// <summary>
+/// Contains the information needed to connect to the backend database.
+/// </summary>
+/// <param name="DatabaseType">Type of database to use.</param>
+/// <param name="ConnectionString">Connection string to access the database.</param>
+/// <param name="Options">Custom options for the specific database. If there are no options this will be empty.</param>
+public record DataSource(DatabaseType DatabaseType, string ConnectionString, Dictionary<string, JsonElement> Options)
+{
+    /// <summary>
+    /// Converts the <c>Options</c> dictionary into a typed options object.
+    /// </summary>
+    /// <typeparam name="TOptionType">The strongly typed object for Options.</typeparam>
+    /// <returns>The strongly typed representation of Options.</returns>
+    /// <exception cref="NotSupportedException">Thrown when the provided <c>TOptionType</c> is not supported for parsing.</exception>
+    public TOptionType GetTypedOptions<TOptionType>() where TOptionType : IDataSourceOptions
+    {
+        HyphenatedNamingPolicy namingPolicy = new();
+
+        if (typeof(TOptionType).IsAssignableFrom(typeof(CosmosDbNoSQLDataSourceOptions)))
+        {
+            return (TOptionType)(object)new CosmosDbNoSQLDataSourceOptions(
+                    Database: ReadStringOption(namingPolicy.ConvertName(nameof(CosmosDbNoSQLDataSourceOptions.Database))),
+                    Container: ReadStringOption(namingPolicy.ConvertName(nameof(CosmosDbNoSQLDataSourceOptions.Container))),
+                    Schema: ReadStringOption(namingPolicy.ConvertName(nameof(CosmosDbNoSQLDataSourceOptions.Schema))),
+                    // The "raw" schema will be provided via the controller to setup config, rather than parsed from the JSON file.
+                    GraphQLSchema: ReadStringOption(namingPolicy.ConvertName(nameof(CosmosDbNoSQLDataSourceOptions.GraphQLSchema))));
+        }
+
+        if (typeof(TOptionType).IsAssignableFrom(typeof(MsSqlOptions)))
+        {
+            return (TOptionType)(object)new MsSqlOptions(
+                SetSessionContext: ReadBoolOption(namingPolicy.ConvertName(nameof(MsSqlOptions.SetSessionContext))));
+        }
+
+        throw new NotSupportedException($"The type {typeof(TOptionType).FullName} is not a supported strongly typed options object");
+    }
+
+    private string? ReadStringOption(string option) => Options.ContainsKey(option) ? Options[option].GetString() : null;
+    private bool ReadBoolOption(string option) => Options.ContainsKey(option) ? Options[option].GetBoolean() : false;
+
+    [JsonIgnore]
+    public string DatabaseTypeNotSupportedMessage => $"The provided database-type value: {DatabaseType} is currently not supported. Please check the configuration file.";
+}
+
+public interface IDataSourceOptions { }
```

Comment:
For my understanding, why is an interface used instead of a parent class?

Reply:
The interface is used to enforce the type parameter on the `GetTypedOptions` method above but doesn't provide any direct functionality itself (since there so far hasn't been any shared options).

I used an interface as it's not something you create directly, but it could be an abstract class if preferred.

Reply:
Thanks for the explanation. I don't have a strong preference on interface vs abstract class. 

I wanted to understand the thought process behind the decision.